### PR TITLE
fix(driver)!: create dedicated events for `umount2` syscall and rename them to "umount2"

### PR DESCRIPTION
**Pull request by Andreagit97**
**What type of PR is this?**

/kind bug

/kind cleanup

**Any specific area of the project related to this PR?**

/area driver-kmod

/area driver-bpf

/area driver-modern-bpf

/area libscap-engine-bpf

/area libscap-engine-kmod

/area libscap-engine-modern-bpf

/area libsinsp

/area tests

**Does this PR require a change in the driver versions?**

No

**What this PR does / why we need it**:

This PR addresses part of the https://github.com/falcosecurity/libs/issues/911 issue.
Before this patch, `umount2` was associated with event pair `PPME_SYSCALL_UMOUNT_E, PPME_SYSCALL_UMOUNT_X` and "umount" name. Every syscall should have its own events pair, with the right name (in this case "umount2"). This PR is the first step in that direction!

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
fix(driver)!: create dedicated events for `umount2` syscall and rename them to "umount2"
```

### CODE REVIEW

**Review by FedeDP — APPROVED** [review]
LGTM!
Thank you!
/approve

**Comment by incertum** [comment]
LGTM!

/hold until we formalize agreement around non-shared event names across the board, see https://github.com/falcosecurity/libs/pull/936#issuecomment-1456350010.

**Comment by Andreagit97** [comment]
I need to rebase it on master, after #937

**Comment by Andreagit97** [comment]
adapted using the new `ppm_sc` table

File: userspace/libscap/linux/scap_ppm_sc.c [hunk 1]
```diff
@@ -420,6 +420,8 @@ static const ppm_sc_code *g_events_to_sc_map[] = {
 	[PPME_SYSCALL_FCHOWNAT_X] = (ppm_sc_code[]){PPM_SC_FCHOWNAT, -1},
 	[PPME_SYSCALL_UMOUNT_1_E] = (ppm_sc_code[]){PPM_SC_UMOUNT, -1},
 	[PPME_SYSCALL_UMOUNT_1_X] = (ppm_sc_code[]){PPM_SC_UMOUNT, -1},
+	[PPME_SYSCALL_UMOUNT2_E] = (ppm_sc_code[]){PPM_SC_UMOUNT2, -1},
```
Comment (FedeDP) [inline userspace/libscap/linux/scap_ppm_sc.c:423]: We should also update `[PPME_SYSCALL_UMOUNT_E,X]` entries to NULL ;)

Comment (FedeDP) [reply userspace/libscap/linux/scap_ppm_sc.c:423]: Given that they are no more sent.

File: userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp [hunk 2]
```diff
@@ -193,13 +193,19 @@ TEST(interesting_syscalls, names_to_sc_set)
...
 	});
 	ASSERT_PPM_SC_CODES_EQ(sc_set_truth, sc_set);
+}
 
+/* This test asserts the behavior of `names_to_sc_set` API when corner cases like `accept/accept4` are involved */
+TEST(interesting_syscalls, names_sc_set_names_corner_cases)
+{
 	static std::unordered_set<std::string> sc_set_names_truth = {"accept",
-	"accept4", "execve", "syncfs", "eventfd", "eventfd2", "umount", "umount2",
+	"accept4", "execve", "syncfs", "eventfd", "eventfd2", "umount",
```
Comment (FedeDP) [inline userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp:202]: I'd add `umount2` in the `tmp_sc_set` isntead of removing it from here.
Or, we could also completely drop `umount2` from here since it is not a corner case anymore.

**Review by FedeDP — COMMENTED** [review]
Found a couple of minor issues, otherwise this LGTM!

File: userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp (hunk 2 above)
Comment (FedeDP) [reply userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp:202]: Interestingly, the test is also failing now:
```
Expected equality of these values:
  std::set<std::string>(sc_set_names_truth.begin(), sc_set_names_truth.end())
    Which is: { "accept", "accept4", "eventfd", "eventfd2", "execve", "pipe", "pipe2", "signalfd", "signalfd4", "syncfs", "umount" }
  std::set<std::string>(sc_set_names.begin(), sc_set_names.end())
    Which is: { "accept", "accept4", "eventfd", "eventfd2", "execve", "pipe", "pipe2", "signalfd", "signalfd4", "syncfs", "umount", "umount2" }
```
Didn't expect that; why is it retrieving "umount2" in the `sc_set_names`?

Comment (FedeDP) [reply userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp:202]: Oh of course, because `[PPME_SYSCALL_UMOUNT_E] = (ppm_sc_code[]){PPM_SC_UMOUNT2, -1},` is mapped to UMOUNT2 here.

File: userspace/libscap/linux/scap_ppm_sc.c (hunk 1 above)
Comment (FedeDP) [reply userspace/libscap/linux/scap_ppm_sc.c:423]: Oh i am wrong, it is an old version and should stay there!

Comment (Andreagit97) [reply userspace/libscap/linux/scap_ppm_sc.c:423]: uhm, I don't think so, `[PPME_SYSCALL_UMOUNT_E,X]` are old versions so it should be correct to associated them to `PPM_SC_UMOUNT2`

File: userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp (hunk 2 above)
Comment (Andreagit97) [reply userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp]: See https://github.com/falcosecurity/libs/pull/944#issuecomment-1460335719

File: userspace/libscap/linux/scap_ppm_sc.c (hunk 1 above)
Comment (Andreagit97) [reply userspace/libscap/linux/scap_ppm_sc.c:423]: In the end, I mapped it to `PPM_SC_UMOUNT`, this entry will be never used and in this way the mapping would be more meaningful `umount_event_pair` -> `ppm_sc_umount`

**Comment by Andreagit97** [comment]
after the rebase the CI is failing on ARM64 i supposed that there is something strange under the hood, i will take a look :eyes:

**Comment by incertum** [comment]
> Hey, folks here there is something strange.. we are using `libsinsp::events::names_to_sc_set()` API to convert `event_names` to `sc_set` while its dual `libsinsp::events::sc_set_to_names` convert `sc_set` to `sc_names` :/ this is quite confusing in my opinion. Right now we can keep it as it is in this PR but my proposal would be:
> 
>     1. create 2 methods that convert `event_names` -> `sc_set` (what is doing `names_to_sc_set`) and vice-versa `sc_set` -> `event_names`
> 
>     2. restore original behavior of `libsinsp::events::names_to_sc_set()`  so `sc_names` -> `sc_set`
> 
>     3. rename all methods like `names_to_sc_set` in something more meaningful like `sc_names_to_sc_set`, "names" alone is too generic IMHO
> 
> 
> @FedeDP @jasondellaluce @incertum

It is the expected current behavior given shared event names across syscalls, e.g. Falco rule calls `accept` -> activate both `accept`, `accept4`. Would propose to hold this PR or any further back and forth until @jasondellaluce formalizes agreement on "requiring Falco to call the real syscall string name in the rules at all times in case of a syscall event". 

Then perhaps we should first have another PR that changes this for other shared event cases, e.g. the `accept` -> activate both `accept`, accept4` case and similar ones. The current workaround in `names_to_sc_set` can then be removed again in that very same PR as it will then be redundant, but it's a big `!` for all Falco end users. 

re  `names_to_sc_set`, I think it is fine to keep the name, but we could rename it (on board either way).

Last step -> rebasing this PR one more time and the confusion and hiccups should have gone away by then.

**Comment by FedeDP** [comment]
> Does this PR require a change in the driver versions?
>
> No

I think we actually need a schema minor version bump. I'm going to do that in #975

**Comment by FedeDP** [comment]
> I think we actually need a schema minor version bump. I'm going to do that in https://github.com/falcosecurity/libs/pull/975

Minor SCHEMA version was already bumped since last release in https://github.com/falcosecurity/libs/commit/dae8bc2f77205679fa6c443b500b72e3d5cc1311

File: userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp [hunk 3]
```diff
@@ -197,14 +197,22 @@ TEST(interesting_syscalls, names_to_sc_set)
...
 	});
 	ASSERT_PPM_SC_CODES_EQ(sc_set_truth, sc_set);
+}
+
+/* This test asserts the behavior of `names_to_sc_set` API when corner cases like `accept/accept4` are involved */
+/// todo: @Andreagit97 revisit this test after new APIs to convert from event_names -> sc_set and sc_set -> event_names.
+TEST(interesting_syscalls, names_sc_set_names_corner_cases)
+{
+	/* INCONSISTENCY: `names_to_sc_set` is converting event names to ppm_sc, but this was not its original scope, the original scope was to convert sc_names -> to sc_set  */
+	std::unordered_set<std::string> event_names{"accept", "execve", "syncfs", "eventfd", "umount", "pipe", "signalfd"};
```
Comment (incertum) [inline userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp:207]: We now have umount2 as event name and system call name so old tests should pass? Would prefer not purging umount2 from the tests.

Agreed on revisiting all tests after all new APIs are finalized.

We already put this on
/hold until after 2023-03-15 community call formal agreement, see https://github.com/falcosecurity/falco/issues/2443

plus need to decide which one to merge first and which one needs subsequent rebasing <-> https://github.com/falcosecurity/libs/pull/975

File: userspace/libsinsp/test/table/event_table.cpp [hunk 4]
```diff
@@ -101,14 +101,59 @@ TEST(event_table, check_unique_events_syscall_category)
...
 			break;
-		}	
+		}
 	}
 
 end:
 	ASSERT_EQ(event_num, PPM_EVENT_MAX);
 }
+
+TEST(events, check_event_names)
```
Comment (Andreagit97) [inline userspace/libsinsp/test/table/event_table.cpp:116]: just to avoid other "accept/accept4" cases in the next future :laughing:

File: userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp (hunk 3 above)
Comment (Andreagit97) [reply userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp:207]: now this test is correctly retrieving only PPM_SC_UMOUNT from the `umount` name, I have to purge "PPM_SC_UMOUNT2" from the expected sc_set otherwise the test won't pass :thinking:

Comment (Andreagit97) [reply userspace/libsinsp/test/public_sinsp_API/interesting_syscalls.cpp]: i've added again `UMOUNT_2` and its associated `PPM_SC_UMOUNT2` and obviously tests are green again :)

**Review by incertum — APPROVED** [review]
Thanks a lot @Andreagit97 !!!

Also quickly tested it with some dummy setup:
```
-- Filter event types names: mount, umount2
-- Start capture
{"evt.dir":">","evt.type":"mount","proc.cmdline":"mount --bind foo2 foo","proc.name":"mount"}
{"evt.dir":"<","evt.type":"mount","proc.cmdline":"mount --bind foo2 foo","proc.name":"mount"}
{"evt.dir":">","evt.type":"umount2","proc.cmdline":"umount foo","proc.name":"umount"}
{"evt.dir":"<","evt.type":"umount2","proc.cmdline":"umount foo","proc.name":"umount"}
{"evt.dir":">","evt.type":"umount2","proc.cmdline":"systemd --user","proc.name":"systemd"}
{"evt.dir":"<","evt.type":"umount2","proc.cmdline":"systemd --user","proc.name":"systemd"}
{"evt.dir":">","evt.type":"umount2","proc.cmdline":"systemd rhgb --switched-root --system --deserialize 35","proc.name":"systemd"}
{"evt.dir":"<","evt.type":"umount2","proc.cmdline":"systemd rhgb --switched-root --system --deserialize 35","proc.name":"systemd"}
```


/approve

/unhold

**Review by FedeDP — APPROVED** [review]
/approve
